lambda/lambda.py

```diff
@@ -0,0 +1,84 @@
+import boto3
+import collections
+import json
+import logging
+import os
+
+import cfnresponse
+
+DEFAULT_LOG_LEVEL = logging.DEBUG
+LOG_LEVELS = collections.defaultdict(
+    lambda: DEFAULT_LOG_LEVEL,
+    {
+        "critical": logging.CRITICAL,
+        "error": logging.ERROR,
+        "warning": logging.WARNING,
+        "info": logging.INFO,
+        "debug": logging.DEBUG,
+    },
+)
+
+# Lambda initializes a root logger that needs to be removed in order to set a
+# different logging config
+root = logging.getLogger()
+if root.handlers:
+    for handler in root.handlers:
+        root.removeHandler(handler)
+
+log_file_name = ""
+if not os.environ.get("AWS_EXECUTION_ENV"):
+    log_file_name = "ami-lookup.log"
+
+logging.basicConfig(
+    filename=log_file_name,
+    format="%(asctime)s.%(msecs)03dZ [%(name)s][%(levelname)-5s]: %(message)s",
+    datefmt="%Y-%m-%dT%H:%M:%S",
+    level=LOG_LEVELS[os.environ.get("LOG_LEVEL", "").lower()],
+)
+log = logging.getLogger(__name__)
+
+ec2 = boto3.client("ec2")
+
+
+def build_search_body(event):
+    owners = json.loads(event["AmiOwners"])
+    filters = [{"Name": "name", "Values": [event["AmiNameSearchString"]]}]
+
+    if event.get("AdditionalFilters"):
+        filters = filters + event["AdditionalFilters"]
+    if event.get("Owners"):
+        owners = event["Owners"]
+
+    return {"Owners": owners, "Filters": filters}
+
+
+def get_ami_id(event):
+    response = ec2.describe_images(**build_search_body(event))
+    amis = sorted(response["Images"], key=lambda x: x["CreationDate"], reverse=True)
+    ami_id = amis[0]["ImageId"]
+
+    log.info("found ami: %s", ami_id)
+    return ami_id
+
+
+def handler(event, context):
+    log.debug("Received event: %s", event)
+
+    response_type = cfnresponse.SUCCESS
+    response_value = event["ResourceProperties"]
+    response_data = {}
+    response_data["Data"] = response_value
+    reason = None
+
+    if  event['RequestType'] != "Delete":
+        try:
+            response_data["Id"] = get_ami_id(response_value)
+            if not response_data["Id"]:
+                response_type = cfnresponse.FAILED
+                reason = "Ami Id search returned no results!"
+        except Exception as e:
```

Comment:
I know bare exceptions are less than ideal but trying to account for everything that can go wrong seems tedious. if this needs to be fleshed out though I can update this.

Reply:
Is there some reason we wouldn't want to just let this raise so the lambda fails? That way the failure is tracked in the lambda metrics...

Reply:
hmm will have to think about that.. I was getting frustrated waiting on cfn to timeout waiting on a response from lambda so ideally i'd want both.. for cfn to be alerted of the failure and for lambda to track it accordingly

Reply:
Well as it is, if there is an exception here, it reports success back to the cfn stack... so need to adjust either way.

```
        try:
            response_data["Id"] = get_ami_id(response_value)
            if not response_data["Id"]:
                raise NoResultsException  # define this as a custom exception
        except Exception:
            cfnresponse.send(<failure>)
            raise
```

Reply:
yeah, realized how simple that was as soon as I hit send..

Reply:
working on the custom exception now

Reply:
updated